### PR TITLE
URLInput: Pass the the search result object to props.onChange

### PR DESCRIPTION
**Pull request by roborourke**
## Description

The `URLInput` component is great however the number of use cases it can support increases a huge amount if it makes the search result object available to the `onChange()` prop as well.

Example use cases:

- Component can be used as a simple post picker
- Retrieving the post title as well to update block attributes eg. link text
- Fetching other / additional post data using the `_links` property of the search result object

## How has this been tested?

I made a copy of the component with the proposed modifications locally and used it as a post picker. I updated not only the URL but was able to set other attributes such as a post title and fetch the post's featured image.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Review by youknowriad — COMMENTED** [review]
LGTM overall, I left some small comments and I'd appreciate other opinions on whether this is a good idea or not. cc @WordPress/gutenberg-core

File: packages/editor/src/components/url-input/index.js
Comment (roborourke) [reply packages/editor/src/components/url-input/index.js]: Not totally sure I follow your meaning, do you mean leave out the default value or change the call to `this.props.onChange()` to `this.props.onChange( ...arguments )` instead? I felt this was a bit more readable and the documentation would be simpler if the type was always consistent.

File: packages/editor/src/components/url-input/index.js
Comment (roborourke) [reply packages/editor/src/components/url-input/index.js]: This is the only place `this.props.onChange()` is called, not passing it here would remove the intended change wouldn't it?

Comment (roborourke) [reply packages/editor/src/components/url-input/index.js]: oh my bad, thought `selectLink()` called `this.onChange()` 🤦‍♂️

File: packages/editor/src/components/url-input/index.js
Comment (youknowriad) [inline packages/editor/src/components/url-input/index.js]: Should we remove the first argument from this method, as it's already in `post.url`

**Review by youknowriad — APPROVED** [review]
LGTM 👍 Anyone for another opinion here?

File: packages/editor/src/components/url-input/README.md
Comment (tofumatt) [inline packages/editor/src/components/url-input/README.md]: What does this mean? Oh, wait, I think I see. I think this should be something like:

> Renders a URL input field; this component should typically be wrapped in a `URLInputButton` component.

right?

File: packages/editor/src/components/url-input/README.md
Comment (tofumatt) [inline packages/editor/src/components/url-input/README.md]: Saying "this is the same" is a bit confusing to me–does that mean it's functionally the same or it's literally passed from `URLInputButton` to this component? If it's just that it behaves similarly I think it's worth duplicating the documentation because it's annoying to have to hunt around a file for docs 😄

File: packages/editor/src/components/url-input/README.md
Comment (tofumatt) [inline packages/editor/src/components/url-input/README.md]: A comma after "rendered" would improve readability here. I also think you could tweak the second sentence:

> By default, the input will gain focus when it is rendered, as typically it is used in combination with a `Popover` in `URLInputBUtton`. If you are not conditionally rendering this component: set this to `false`.

**Review by tofumatt — CHANGES_REQUESTED** [review]
I think this is a solid change but there are documentation tweaks I think are worth doing before merging. I can make them in a few hours if no one can get to them beforehand–if they're made before I get the chance I'll re-review and merge when I'm back 😄

File: packages/editor/src/components/url-input/README.md
Comment (roborourke) [reply packages/editor/src/components/url-input/README.md]: It's literally passed directly from the `URLInputButton` component. But I think there's no harm in duplicating it for the sake of ease as you say 👍

File: packages/editor/src/components/url-input/README.md
Comment (tofumatt) [reply packages/editor/src/components/url-input/README.md]: Agreed, that's great 👍

**Review by tofumatt — APPROVED** [review]

**Comment by tofumatt** [comment]
Changes are good; thanks a bunch!

**Comment by roborourke** [comment]
@tofumatt @youknowriad thank you!

**Pull request merged**
